### PR TITLE
adds sentiment

### PR DESCRIPTION
**Pull request by JeroenJansze2018**

### CODE REVIEW

File: textpipe/doc.py [hunk 1]
```diff
@@ -274,3 +275,20 @@ def _set_default_nlp(self, lang):
             self._spacy_nlps[lang] = {}
         model = spacy.load('{}_core_{}_sm'.format(lang, 'web' if lang == 'en' else 'news'))
         self._spacy_nlps[lang][None] = model
+
+    @functools.lru_cache()
+    def sentiment(self, lang):
```
Comment (anneschuth) [inline textpipe/doc.py]: there's `self.language`

File: textpipe/operation.py [hunk 2]
```diff
@@ -109,10 +115,31 @@ class Entities(Operation):
...
+    >>> doc = Doc('Een hele leuke test zin.')
+    >>> Sentiment()(doc)
+    (0.9599999999999999, 1.0)
+    """
+
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+
+    def __call__(self, doc):
+        lang = doc.language if doc.is_reliable_language else doc.hint_language
```
Comment (anneschuth) [inline textpipe/operation.py]: I think this logic can move to the doc itself

File: textpipe/doc.py [hunk 3]
```diff
@@ -274,3 +275,20 @@ def _set_default_nlp(self, lang):
...
+
+        >>> doc = Doc('Een leuke test zin.')
+        >>> doc.sentiment('nl')
+        (0.6, 0.9666666666666667)
+        """
+
+        if lang == 'nl':
+            return sentiment(self.clean_text())
+        else:
+            return 0.0, 0.0
```
Comment (anneschuth) [inline textpipe/doc.py]: I think a not-implemented error should be raised instead of silently returning `0`

File: textpipe/doc.py [hunk 4]
```diff
@@ -274,3 +275,20 @@ def _set_default_nlp(self, lang):
...
+        Returns sentiment score (-1 to 1) and a confidence level (0 to 1)
+
+        Currently only Dutch supported
+
+        >>> doc = Doc('Een leuke test zin.')
+        >>> doc.sentiment('nl')
+        (0.6, 0.9666666666666667)
+        """
+
+        if lang == 'nl':
```
Comment (anneschuth) [inline textpipe/doc.py]: maybe we should support all languages supported by pattern?

File: textpipe/doc.py [hunk 5]
```diff
@@ -274,3 +275,20 @@ def _set_default_nlp(self, lang):
             self._spacy_nlps[lang] = {}
         model = spacy.load('{}_core_{}_sm'.format(lang, 'web' if lang == 'en' else 'news'))
         self._spacy_nlps[lang][None] = model
+
+    @functools.lru_cache()
+    def sentiment(self):
```
Comment (anneschuth) [inline textpipe/doc.py]: this can now be a property

Comment (JeroenJansze2018) [reply textpipe/doc.py]: Does not seem to be working:

Failed example:
    doc.sentiment()
Exception raised:
    Traceback (most recent call last):
      File "/snap/pycharm-community/85/helpers/pycharm/docrunner.py", line 140, in __run
        compileflags, 1), test.globs)
      File "<doctest doc.Doc.sentiment[1]>", line 1, in <module>
        doc.sentiment()
    TypeError: 'Score' object is not callable

I take it that Score is something from pattern

Comment (anneschuth) [reply textpipe/doc.py]: you should remove the `()` when 'calling` a property (from the operation, for instance)

File: textpipe/doc.py [hunk 6]
```diff
@@ -274,3 +275,20 @@ def _set_default_nlp(self, lang):
...
+
+        Currently only Dutch supported
+
+        >>> doc = Doc('Een leuke test zin.')
+        >>> doc.sentiment
+        (0.6, 0.9666666666666667)
+        """
+
+        if self.language == 'nl':
+            return sentiment(self.clean_text())
```
Comment (anneschuth) [inline textpipe/doc.py]: I think we should use `self.clean` instead.

File: textpipe/doc.py (hunk 4 above)
Comment (JeroenJansze2018) [reply textpipe/doc.py]: This is what I had at first. But english gives always (0.0, 0.0)

File: textpipe/doc.py (hunk 6 above)
Comment (msappelli) [reply textpipe/doc.py]: agree, although I see that in _load_spacy_doc clean_text() is used as well..

File: textpipe/doc.py [hunk 7]
```diff
@@ -276,15 +276,15 @@ def _set_default_nlp(self, lang):
         model = spacy.load('{}_core_{}_sm'.format(lang, 'web' if lang == 'en' else 'news'))
         self._spacy_nlps[lang][None] = model
 
-    @functools.lru_cache()
+    @property
     def sentiment(self):
         """
         Returns sentiment score (-1 to 1) and a confidence level (0 to 1)
```
Comment (msappelli) [inline textpipe/doc.py]: Please update comments - see documentation of pattern, pattern returns a tuple of polarity and subjectivity score - this is not the same as sentiment & confidence!
https://www.clips.uantwerpen.be/pages/pattern-nl

File: textpipe/doc.py [hunk 8]
```diff
@@ -274,3 +277,26 @@ def _set_default_nlp(self, lang):
...
+
+        if self.language == 'en':
+            return sentiment_en(self.clean)
+        elif self.language == 'nl':
+            return sentiment_nl(self.clean)
+        elif self.language == 'fr':
+            return sentiment_fr(self.clean)
+        elif self.language == 'it':
+            return sentiment_it(self.clean)
+        else:
```
Comment (anneschuth) [inline textpipe/doc.py]: no need for the `else`

File: textpipe/doc.py [hunk 9]
```diff
@@ -10,7 +10,10 @@
 import textacy
 import textacy.text_utils
 from bs4 import BeautifulSoup
-
+from pattern.text.nl import sentiment as sentiment_nl
+from pattern.text.en import sentiment as sentiment_en
+from pattern.text.fr import sentiment as sentiment_fr
+from pattern.text.it import sentiment as sentiment_it
```
Comment (graus) [inline textpipe/doc.py]: Now we load all languages in memory, regardless of which you'll use/need. 
Not sure how to make this more lazy, but that would be nice. Any thoughts on it?

On second look, it does look like it is loaded lazily; 

```
In [1]: from pattern.text.en import sentiment as sentiment_en

In [2]: import sys

In [3]: sys.getsizeof(sentiment_en)
Out[3]: 256
[...] -> print(sentiment_en)
In [5]: sys.getsizeof(sentiment_en)
Out[5]: 147576
```

Comment (graus) [reply textpipe/doc.py]: It takes ~4sec to load the EN object;
```
In [3]: t0=time();from pattern.text.en import sentiment as sentiment_en;print(time()-t0)
3.775541067123413
```
The other languages come for free;
```
In [4]: t0=time();from pattern.text.en import sentiment as sentiment_fr;print(time()-t0)
6.103515625e-05

In [5]: t0=time();from pattern.text.en import sentiment as sentiment_nl;print(time()-t0)
6.198883056640625e-05

In [6]: t0=time();from pattern.text.en import sentiment as sentiment_de;print(time()-t0)
7.104873657226562e-05
```

**Pull request merged**